File: core/src/main/java/com/simple/castle/scene/game/controller/Player.java
package com.simple.castle.scene.game.controller;

import com.badlogic.gdx.math.Vector3;
import com.simple.castle.core.object.constructors.ObjectConstructors;
import com.simple.castle.core.object.unit.abs.AbstractGameObject;

import java.util.*;
import java.util.stream.Collectors;

public class Player {
    private final Vector3 tempVector = new Vector3();
    private final String playerName;
    private final String unitType;
    private final Set<PlayerUnit> units = new HashSet<>();

    private final List<List<AbstractGameObject>> paths;
    private final List<Vector3> initPositions;

    public Player(String unitType, List<List<AbstractGameObject>> paths, String playerName) {
        this.unitType = unitType;
        this.paths = paths;
        this.initPositions = paths.stream()
                .map(path -> path.stream().findFirst())
                .filter(Optional::isPresent)
                .map(Optional::get)
                .map(sceneObject -> sceneObject.transform.getTranslation(tempVector).cpy())
                .collect(Collectors.toList());
        this.playerName = playerName;
    }

    private static AbstractGameObject getNextAvailable(List<AbstractGameObject> list, AbstractGameObject current) {
        int i = list.indexOf(current);
        if (i >= 0 && i < list.size() - 1) {
            i += 1;
        }
        return list.get(i);
    }

    public void update() {
        units.forEach(PlayerUnit::update);
    }

    public void collisionEvent(PlayerUnit playersUnit, AbstractGameObject anotherObject) {
        if (playersUnit != null && anotherObject != null) {
            paths.stream()
                    .filter(path -> path.contains(anotherObject))
                    .findAny()
                    .map(path -> getNextAvailable(path, anotherObject))
                    .ifPresent(path -> playersUnit.setMovePoint(path.transform.getTranslation(tempVector).cpy()));
            if (anotherObject instanceof PlayerUnit && !isPlayers((PlayerUnit) anotherObject)) {
                playersUnit.setDeath(true);
            }
        }
    }

    public List<PlayerUnit> spawnUnitsOnStartPositions(ObjectConstructors objectConstructors) {
        return initPositions.stream()
                .map(initPosition ->
                        new PlayerUnit(objectConstructors.getConstructor(unitType), initPosition, playerName))
                .peek(units::add)
                .collect(Collectors.toList());
    }

    public boolean isPlayers(PlayerUnit gameObject) {
        return units.contains(gameObject);
    }

    public Collection<PlayerUnit> getUnits() {
        return units;
    }

    public Collection<PlayerUnit> getDeadUnitsAndClear() {
        Set<PlayerUnit> deadUnits = units.stream().filter(PlayerUnit::isDead).collect(Collectors.toSet());
        units.removeAll(deadUnits);
        return deadUnits;
    }
}

File: server/server-core/src/main/java/com/simple/castle/server/kt/controller/Player.kt
package com.simple.castle.server.kt.controller

import com.badlogic.gdx.math.Vector3
import com.simple.castle.server.kt.composition.BaseObject
import com.simple.castle.server.kt.manager.SceneManager
import com.simple.castle.server.kt.physic.PhysicWorld

class Player(private val playerName: String,
             private val unitType: String,
             private val path: List<List<BaseObject>>,
             private val startPositions: List<Vector3>,
             private val sceneManager: SceneManager,
             private val physicWorld: PhysicWorld) {
    private val units: MutableMap<String, PlayerUnit> = HashMap()

    fun spawnOnStartPositions() {
        startPositions
                .map { startPosition -> PlayerUnit(sceneManager.getConstructor(unitType)!!, startPosition) }
                .onEach { playerUnit -> units[playerUnit.id] = playerUnit }
                .forEach { playerUnit ->
                    run {
                        sceneManager.addObjects(playerUnit)
                        physicWorld.addRigidBody(playerUnit.physicObject)
                    }
                }
    }

    fun update() {
        units.values.forEach { playerUnit -> playerUnit.update() }
    }

    fun onContactStarted(baseObject1: BaseObject, baseObject2: BaseObject) {
        if (baseObject1.id in units) {
            move(baseObject1 as PlayerUnit, baseObject2)
        }

        if (baseObject2.id in units) {
            move(baseObject2 as PlayerUnit, baseObject1)
        }
    }

    private fun move(unit: PlayerUnit, area: BaseObject) {
        if (area in path.flatten()) {
            val nextPoint = path.flatten()[path.flatten().indexOf(area) + 1]
            unit.setMovePoint(nextPoint.modelInstance.transform.getTranslation(Vector3()))
        }
    }
}